File: 09b.pl
#! /usr/bin/perl
use warnings;
use strict;
use feature qw{ say };

use ARGV::OrDATA;

my %DIR = (R => [1, 0],
           L => [-1, 0],
           U => [0, -1],
           D => [0, 1]);

my @rope = map [0, 0], 1 .. 10;
my %grid;
$grid{0}{0} = 1;
my $count = 1;

while (<>) {
    my ($direction, $length) = split;
    for (1 .. $length) {
        $rope[0][0] += $DIR{$direction}[0];
        $rope[0][1] += $DIR{$direction}[1];

        for my $tail (1 .. 9) {
            my ($hx, $hy, $tx, $ty)
                = (@{ $rope[ $tail - 1 ] }, @{ $rope[$tail] });
            my $dx = $hx - $tx;
            my $dy = $hy - $ty;
            next unless abs($dx) > 1 || abs($dy) > 1;

            $tx += $dx > 0 ? 1 : -1 if $dx;
            $ty += $dy > 0 ? 1 : -1 if $dy;
            $rope[$tail] = [$tx, $ty];
        }
        ++$count unless $grid{ $rope[9][0] }{ $rope[9][1] };
        $grid{ $rope[9][0] }{ $rope[9][1] } = 1;
    }
}

say $count;

__DATA__
R 5
U 8
L 8
D 3
R 17
D 10
L 25
U 20

File: 09b.py
#! /usr/bin/python3
import sys

DIR = {'R': (1, 0),
       'L': (-1, 0),
       'U': (0, -1),
       'D': (0, 1)}

rope = list(map(lambda _x: [0, 0], range(0, 10)))
grid = {0: {0: 1}}
count = 1

with open(sys.argv[1], 'r') as fh:
    for line in fh:
        direction, length = line.split()
        for i in range(1, 1 + int(length)):
            rope[0][0] += DIR[direction][0]
            rope[0][1] += DIR[direction][1]

            for tail in range(1, 10):
                hx, hy, tx, ty = *rope[tail - 1], *rope[tail]
                dx = hx - tx
                dy = hy - ty
                if abs(dx) <= 1 and abs(dy) <= 1:
                    continue

                if dx:
                    tx += 1 if dx > 0 else -1
                if dy:
                    ty += 1 if dy > 0 else -1
                rope[tail] = [tx, ty]

            tx, ty = rope[9]
            if not tx in grid:
                grid[tx] = dict()
            if not ty in grid[tx]:
                count += 1
                grid[tx][ty] = 1
print(count)
